cli/fullservice.py

```diff
@@ -0,0 +1,1173 @@
+    root.write(base64.b64decode(utils.get_secret("ROOTCRT")))
+    root.flush()
+    client = open("client.full.pem", "wb")
+    client.write(base64.b64decode(utils.get_secret("CLIENTCRT")))
+    client.flush()
+
+    ssl_context.load_verify_locations("rootcrt.pem")
+    ssl_context.verify_mode = ssl.CERT_REQUIRED
+    ssl_context.load_cert_chain(certfile="client.full.pem")
+
+
+class Request:
+    url = utils.get_secret('URL')
+    async def req(self, request_data: dict) -> dict:
+        logging.info("request: %s", request_data.get("method"))
+        request_data["params"] = {
+            k: v for k, v in request_data["params"].items() if v
+        }  # handle optional params
+        response_data = await self.request(request_data)
+        if "error" in str(response_data):
+            logging.error(response_data)
+        else:
+            print(response_data)
+        return response_data
+
+    async def request(self, request_data: dict):
+        request_data = {"jsonrpc": "2.0", "id": "1", **request_data}
+        print(f"request data: {request_data}")
+        async with aiohttp.TCPConnector(ssl=ssl_context) as conn:
+            async with aiohttp.ClientSession(connector=conn) as sess:
+                # this can hang (forever?) if there's no full-service at that url
+                async with sess.post(
+                    self.url,
+                    data=json.dumps(request_data),
+                    headers={"Content-Type": "application/json"},
+                ) as resp:
+                    # print(resp.json)
+                    return await resp.json()
+
+
+class FullServiceAPIv2(Request):
+    async def assign_address_for_account(self, account_id, metadata=""):
+        return await self.req(
+            {
+                "method": "assign_address_for_account",
+                "params": {"account_id": account_id, "metadata": metadata},
+            }
+        )
+
+    async def build_and_submit_transaction(
+        self,
+        account_id,
+        addresses_and_amounts="",
+        recipient_public_address="",
+        amount="",
+        input_txo_ids="",
+        fee_value="",
+        fee_token_id="",
+        tombstone_block="",
+        max_spendable_value="",
+        comment="",
+    ):
+        return await self.req(
+            {
+                "method": "build_and_submit_transaction",
+                "params": {
+                    "account_id": account_id,
+                    "addresses_and_amounts": addresses_and_amounts,
+                    "recipient_public_address": recipient_public_address,
+                    "amount": amount,
+                    "input_txo_ids": input_txo_ids,
+                    "fee_value": fee_value,
+                    "fee_token_id": fee_token_id,
+                    "tombstone_block": tombstone_block,
+                    "max_spendable_value": max_spendable_value,
+                    "comment": comment,
+                },
+            }
+        )
+
```

Comment:
what happens when these aren't all set?

Reply:
should be handled on line 38-39 in the req function, it rebuilds the dictionary every request